### PR TITLE
Ee 16064 candidates for archive

### PR DESCRIPTION
**Pull request by mikehalmaho**
Add an orchestrator (ArchiveAuditService) to retrieve audit history, calculate results per nino, and submit each nino back to the audit service as a candidate for archiving.

### CODE REVIEW

File: src/main/java/uk/gov/digital/ho/proving/income/audit/AuditClient.java [hunk 1]
```diff
@@ -65,6 +76,18 @@ private void dispatchAuditableData(AuditableData auditableData) {
         restTemplate.exchange(auditEndpoint, POST, toEntity(auditableData), Void.class);
     }
 
+    List<AuditRecord> getAuditHistory(LocalDate toDate, List<AuditEventType> eventTypes) {
+        AuditHistoryRequest request = new AuditHistoryRequest(toDate, eventTypes);
+        HttpEntity<AuditHistoryRequest> entity = new HttpEntity<>(request, generateRestHeaders());
+        ResponseEntity<List<AuditRecord>> auditRecords = restTemplate.exchange(auditHistoryEndpoint, POST, entity, new ParameterizedTypeReference<List<AuditRecord>>() {});
```
Comment (owainrobertsbjss) [inline src/main/java/uk/gov/digital/ho/proving/income/audit/AuditClient.java]: As the request has nothing sensitive in it, should we be more RESTful and use GET?

File: src/main/java/uk/gov/digital/ho/proving/income/audit/AuditArchiveService.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        List<AuditResult> byCorrelationId = auditResultConsolidator.auditResultsByCorrelationId(auditHistory);
+        List<AuditResultByNino> byNino = auditResultConsolidator.auditResultsByNino(byCorrelationId);
+
+        for (AuditResultByNino auditResult : byNino) {
+            auditClient.archiveAudit(generateAuditHistoryRequest(auditResult, config));
+        }
+    }
+
+    private LocalDate getLastDayToBeArchived() {
+        return LocalDate.now().minusMonths(retainAuditHistoryMonths).minusDays(1);
```
Comment (owainrobertsbjss) [inline src/main/java/uk/gov/digital/ho/proving/income/audit/AuditArchiveService.java:45]: I think this method should be tested independently to make sure it does what we expect, especially around leap years.

File: src/main/java/uk/gov/digital/ho/proving/income/audit/AuditResultConsolidator.java [hunk 3]
```diff
@@ -34,7 +37,26 @@ public AuditResultConsolidator(
...
+    }
+
+    private AuditResultByNino consolidate(List<AuditResult> results) {
+        AuditResult consolidatedResult = results.stream()
+            .max(auditResultComparator)
+            .orElse(null);
+        List<String> allCorrelationIds = results.stream()
+            .map(AuditResult::correlationId)
+            .collect(Collectors.toList());
+        return new AuditResultByNino(consolidatedResult.nino(), allCorrelationIds, consolidatedResult.date(), consolidatedResult.resultType());
```
Comment (owainrobertsbjss) [inline src/main/java/uk/gov/digital/ho/proving/income/audit/AuditResultConsolidator.java:61]: IntelliJ complains that 'consolidatedResult.nino()' could throw a NullPointerException, presumably due to line 52. I'm convinced that it won't be a problem but it's a shame that the Optional has caused this clunkyness. 

I'll think about a possible improvement but we may have to live with it.

File: src/test/java/uk/gov/digital/ho/proving/income/audit/AuditArchiveServiceIT.java [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+            .andExpect(jsonPath("$.eventIds.*", containsInAnyOrder("corr-id-2")))
+            .andExpect(jsonPath("$.result", is("PASS")))
+            .andExpect(jsonPath("$.resultDate", is("2019-01-01")))
+            .andRespond(withSuccess());
+
+        auditArchiveService.archiveAudit();
+    }
+
+    @Test
+    public void archiveAudit_noResponse_errorReturned() {
```
Comment (owainrobertsbjss) [inline src/test/java/uk/gov/digital/ho/proving/income/audit/AuditArchiveServiceIT.java]: It took me a while to work out what this test does and I had to step through it in the debugger to do so. Consider changing the name to reflect that the Mock Audit Service is returning a malformed response (or at least one without a PASS/FAIL/NOT_FOUND in).

File: src/test/java/uk/gov/digital/ho/proving/income/audit/AuditArchiveServiceIT.java [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+            .expect(requestTo(containsString("/history")))
+            .andExpect(method(POST))
+            .andRespond(withSuccess(auditHistory, APPLICATION_JSON));
+
+        String endOfArchive = LocalDate.now().minusMonths(6).minusDays(1).format(DateTimeFormatter.ISO_DATE);
+        mockAuditService
+            .expect(requestTo(containsString("/archive")))
+            .andExpect(method(POST))
+            .andExpect(jsonPath("$.nino", is("nino_1")))
+            .andExpect(jsonPath("$.lastArchiveDate", is(endOfArchive)))
```
Comment (owainrobertsbjss) [inline src/test/java/uk/gov/digital/ho/proving/income/audit/AuditArchiveServiceIT.java]: I don't like that 'is' matcher as it looks more like an identity check than an equality one.

File: src/test/java/uk/gov/digital/ho/proving/income/audit/AuditArchiveServiceIT.java [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+            .andExpect(jsonPath("$.nino", is("nino_1")))
+            .andExpect(jsonPath("$.eventIds.*", containsInAnyOrder("corr-id-1", "corr-id-2")))
+            .andExpect(jsonPath("$.result", is("NOTFOUND")))
+            .andRespond(withSuccess());
+
+        auditArchiveService.archiveAudit();
+    }
+
+    @Test
+    public void archiveAudit_multipleMatchingResults_latestResultReturned() {
```
Comment (owainrobertsbjss) [inline src/test/java/uk/gov/digital/ho/proving/income/audit/AuditArchiveServiceIT.java:137]: Hopefully you can point it out to me but I cannot see how this test asserts that the latest result is returned.

File: src/test/java/uk/gov/digital/ho/proving/income/audit/AuditArchiveServiceIT.java [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+            .andExpect(method(POST))
+            .andExpect(jsonPath("$.nino", is("nino_1")))
+            .andExpect(jsonPath("$.lastArchiveDate", is(endOfArchive)))
+            .andRespond(withSuccess());
+
+        auditArchiveService.archiveAudit();
+    }
+
+    @Test
+    public void archiveAudit_multipleMatchingAndMultipleresults_latestResultForAllRequestsReturned() {
```
Comment (owainrobertsbjss) [inline src/test/java/uk/gov/digital/ho/proving/income/audit/AuditArchiveServiceIT.java]: results should be Results

Comment (owainrobertsbjss) [inline src/test/java/uk/gov/digital/ho/proving/income/audit/AuditArchiveServiceIT.java]: This test is too big for me to understand. Wouldn't 2 different Ninos be enough to confirm it works for multiple? And the PASS/FAIL/NOT_FOUND precedence can be done in other, smaller tests.

File: src/test/java/uk/gov/digital/ho/proving/income/audit/AuditClientTest.java [hunk 8]
```diff
@@ -106,4 +106,33 @@ public void shouldSetAuditableData() {
...
+    @Test
+    public void shouldRetrieveAuditHistory() {
+        List<AuditEventType> eventTypes = Arrays.asList(INCOME_PROVING_FINANCIAL_STATUS_REQUEST, INCOME_PROVING_FINANCIAL_STATUS_RESPONSE);
+        List<AuditRecord> results = new ArrayList<>();
+        ResponseEntity<List<AuditRecord>> resultsEntity = ResponseEntity.ok(results);
+        when(mockRestTemplate.exchange(eq("some history endpoint"), eq(POST), captorHttpEntity.capture(), eq(new ParameterizedTypeReference<List<AuditRecord>>() {}))).thenReturn(resultsEntity);
+
+        List<AuditRecord> auditRecords = auditClient.getAuditHistory(LocalDate.now(), eventTypes);
+
+        verify(mockRestTemplate).exchange(eq("some history endpoint"), eq(POST), captorHttpEntity.capture(), eq(new ParameterizedTypeReference<List<AuditRecord>>() {}));
```
Comment (owainrobertsbjss) [inline src/test/java/uk/gov/digital/ho/proving/income/audit/AuditClientTest.java]: As you're asserting against the captured argument, there's no need to verify the mock.

File: src/test/java/uk/gov/digital/ho/proving/income/audit/AuditClientTest.java [hunk 9]
```diff
@@ -106,4 +106,33 @@ public void shouldSetAuditableData() {
         assertThat(auditableData.getEventType()).isEqualTo(INCOME_PROVING_FINANCIAL_STATUS_REQUEST);
         assertThat(auditableData.getData()).isEqualTo("{}");
     }
+
+    @Test
+    public void shouldRetrieveAuditHistory() {
+        List<AuditEventType> eventTypes = Arrays.asList(INCOME_PROVING_FINANCIAL_STATUS_REQUEST, INCOME_PROVING_FINANCIAL_STATUS_RESPONSE);
+        List<AuditRecord> results = new ArrayList<>();
+        ResponseEntity<List<AuditRecord>> resultsEntity = ResponseEntity.ok(results);
+        when(mockRestTemplate.exchange(eq("some history endpoint"), eq(POST), captorHttpEntity.capture(), eq(new ParameterizedTypeReference<List<AuditRecord>>() {}))).thenReturn(resultsEntity);
```
Comment (owainrobertsbjss) [inline src/test/java/uk/gov/digital/ho/proving/income/audit/AuditClientTest.java]: As the endpoint names have to match the ones passed into AuditClient in the setUp method for this to work it would be nice if there was a constant or a field equal to "some history endpoint"

File: src/test/java/uk/gov/digital/ho/proving/income/audit/AuditResultConsolidatorIT.java [hunk 10]
```diff
@@ -132,19 +138,146 @@ public void byCorrelationId_notFoundResponseOnly_allDetailsFilled() {
...
+        AuditRecord auditRecord2Request = fileUtils.buildRequestRecord("5e6d002f-fd09-4347-a7da-2cd23346da49", "2019-02-26 12:01:02.003", "PP151005D");
+        AuditRecord auditRecord2Response = fileUtils.buildResponseRecord("5e6d002f-fd09-4347-a7da-2cd23346da49", "2019-02-26 12:01:02.003", "PP151005D", "true");
+        List<AuditRecord> records = Arrays.asList(auditRecord1Request, auditRecord1Response, auditRecord2Request, auditRecord2Response);
+
+        List<AuditResult> results = auditResultConsolidator.auditResultsByCorrelationId(records);
+
+        AuditResult expected1 = getExpectedAuditResult1();
+        AuditResult expected2 = getExpectedAuditResult2();
+
+        assertThat(results.size()).isEqualTo(2);
```
Comment (owainrobertsbjss) [inline src/test/java/uk/gov/digital/ho/proving/income/audit/AuditResultConsolidatorIT.java]: containsExactlyInAnyOrder will fail if there's not 2 elements, so this line can go.

**Review by owainrobertsbjss — CHANGES_REQUESTED** [review]
I think the uk.gov.digital.ho.proving.income.audit package might need some splitting as it's got 16 classes in it with very similar names.

File: src/main/java/uk/gov/digital/ho/proving/income/audit/AuditArchiveService.java (hunk 2 above)
Comment (mikehalmaho) [reply src/main/java/uk/gov/digital/ho/proving/income/audit/AuditArchiveService.java:45]: As discussed the date is arbitrary.

File: src/main/java/uk/gov/digital/ho/proving/income/audit/AuditResultConsolidator.java (hunk 3 above)
Comment (mikehalmaho) [reply src/main/java/uk/gov/digital/ho/proving/income/audit/AuditResultConsolidator.java:61]: Agreed, and I also need to add a test case for the empty list scenario.

File: src/test/java/uk/gov/digital/ho/proving/income/audit/AuditArchiveServiceIT.java (hunk 4 above)
Comment (mikehalmaho) [reply src/test/java/uk/gov/digital/ho/proving/income/audit/AuditArchiveServiceIT.java]: OK this needs a better method name.

File: src/test/java/uk/gov/digital/ho/proving/income/audit/AuditArchiveServiceIT.java (hunk 5 above)
Comment (mikehalmaho) [reply src/test/java/uk/gov/digital/ho/proving/income/audit/AuditArchiveServiceIT.java]: First I'll see if I can use AssertJ as we do in the tests - if not then I'll find a better method name.

File: src/test/java/uk/gov/digital/ho/proving/income/audit/AuditArchiveServiceIT.java (hunk 7 above)
Comment (mikehalmaho) [reply src/test/java/uk/gov/digital/ho/proving/income/audit/AuditArchiveServiceIT.java]: Let me have a quick refactor and see what you think - if it's still too big then I'll have a rethink.

**Review by owainrobertsbjss — APPROVED** [review]
